src/EventSubscriber/IndexEventSubscriber.php

```diff
@@ -113,21 +114,25 @@ public function onIndexEvent(IndexReadyEvent $event): void
                                 ->setSource($source);
                         }
                     }
-                }
 
-                // Make every thing stick.
-                $this->em->flush();
-                $this->em->getConnection()->commit();
+                    $this->em->flush();
+                }
+            } catch (UniqueConstraintViolationException $exception) {
+                // Some vendors have more than one unique identifier in the input data, so to queue processors can try
+                // to write the same row to the search table, which is not allowed. So we restart the entity manager to
+                // ensure that it can write the last index timestamp to the database. If we do not do this we may
+                // end up in a loop of the same records falling again and again.
+                $this->metricsService->counter('index_event_unique_violation', 'Index event unique constraint violation', 1, ['type' => 'index']);
+                $this->registry->resetManager();
```

Comment:
Please note the following from the `resetManager` doc comment:
```
     * Be warned that you will get a brand new object manager as
     * the existing one is not useable anymore. This means that any
     * other object with a dependency on this object manager will
     * hold an obsolete reference. You can inject the registry instead
     * to avoid this problem.
```
So I think we need to do:
```php
$this->em = $this->registry->resetManager();
```